File: LeetCode/0279. Perfect Squares/solution.cpp
/**
 * 588 / 588 test cases passed.
 * Runtime: 220 ms
 * Memory Usage: 9 MB 
 */
class Solution {
public:
    int numSquares(int n) {
        vector<int> dp(n + 1, INT_MAX);
        dp[0] = 0;
        for (int i = 1; i <= n; ++ i) {
            int x = i * i;
            for (int j = x; j <= n; ++ j) {
                dp[j] = min(dp[j], dp[j - x] + 1);
            }
        }
        return dp[n];
    }
};

File: LeetCode/0279. Perfect Squares/solution.py
"""
588 / 588 test cases passed.
Runtime: 3848 ms
Memory Usage: 15.3 MB
"""
class Solution:
    def numSquares(self, n: int) -> int:
        dp = list(range(n + 1))
        for i in range(1, n):
            if i * i > n: break
            for j in range(i * i, n + 1):
               dp[j] = min(dp[j], dp[j - i * i] + 1)
        return dp[-1]
